research/02-repl.tex

```diff
@@ -1,5 +1,172 @@
+\textit{REPL}. There are subtle differences between these interpretations.
+These are, however, mostly of semantic value. In this report the term
+REPL is chosen, because it conveys the notion of such an interactive
+environment well.
+
+The Lisp family of programming languages is believed to be amongst the
+first programming languages offering such functionality~\cite{NEEDED}. The name
+REPL comes from the Lisp functions that implement a REPL's functionality:
+
+\begin{enumerate}
+  \item The \texttt{read} function takes a user's input, which often is just one
+    or several expressions as opposed to a complete compilation unit. It then
+    parses this input and creates an abstract syntax tree (AST).
+  \item The AST created in the previous step is then passed on to the
+    \texttt{eval} function, which evaluates it.
+  \item The result yielded by the previous function is then printed out to the
+    user by the \texttt{print} function.
+  \item After having printed the result, the environment needs to \texttt{loop}
+    back to the read state.
+\end{enumerate}
+
+Assuming the individual functions listed previously exist, a REPL can be created
+in a single line of code simply by combining the functions:
+
+\begin{lstlisting}
+(loop (print (eval (read))))
+\end{lstlisting}
+
+Part of the semantic differences between the different names for an interactive
+programming environment is Lisp's homoiconicity. In Lisp, the structure of
+expressions is represented directly in a data structure, resulting in the
+ability to infer a program's or data object's state simply by reading its
+textual representation. In Lisp REPLs, therefore, arbitrary data objects yielded
+from a previous expression can be used as input to the next expression.
+Homoiconicity is an unusual feature in modern programming languages not
+belonging to the Lisp family, however, so interactive environments of these
+programming languages often require additional hoops to read and evaluate
+expressions. As said earlier, this difference is mostly semantic and therefore
+this report sticks with the name \textit{REPL}.
+
+REPLs offer several advantages that make them an often requested feature.
+Because of their nature, REPLs provide the ability to program interactively.
+%Interactive programming is a programming paradigm in which parts of a program
+%are written while it is already active~\cite{NEEDED}.
+Programming interactively has multiple applications.
+
+% TODO: mention REPL Driven Development (RDD)?
+When creating software solutions for an as of yet not well understood domain,
+often it is not clear which data structures and algorithms are required. In such
+cases, REPLs offer the ability to interactively develop and debug software
+without having to apply the (oftentimes much slower) edit-compile-run-debug
+development style. This kind of programming is called exploratory
+programming~\cite{NEEDED}. Related to this kind of exploration, REPLs also
```

Comment:
I think I could take one or two references from [this website](http://perchta.fit.vutbr.cz/projekty/22). I am as of yet unable to find a proper academic paper on exploratory programming.